relay-server/src/extractors/decoder.rs

```diff
@@ -0,0 +1,212 @@
+    ///
+    /// If the request is not encoded, this decoder is a noop.
+    pub fn new<S>(request: &HttpRequest<S>, limit: usize) -> Self {
+        let sink = Sink::new(limit);
+
+        let encoding = request
+            .headers()
+            .get(CONTENT_ENCODING)
+            .and_then(|enc| enc.to_str().ok())
+            .map(HttpEncoding::parse)
+            .unwrap_or_default();
+
+        let inner = match encoding {
+            HttpEncoding::Identity => DecoderInner::Identity(Box::new(sink)),
+            HttpEncoding::Br => DecoderInner::Br(Box::new(BrotliDecoder::new(sink))),
+            HttpEncoding::Gzip => DecoderInner::Gzip(Box::new(GzDecoder::new(sink))),
+            HttpEncoding::Deflate => DecoderInner::Deflate(Box::new(DeflateDecoder::new(sink))),
+        };
+
+        Self { inner }
+    }
+
+    /// Decodes a chunk of data.
+    ///
+    /// The decoded bytes are written into the Decoder's buffer, which can be obtained using
+    /// [`take`](Self::take). Returns `Ok(false)` if decoding has completed successfully without an
+    /// overflow. Returns `Ok(true)` if decoding has stopped prematurely due to an overflow. In this
+    /// case, the buffer contains the decoded payload up to the limit. Returns `Err` if there was an
+    /// error decoding.
+    pub fn decode(&mut self, bytes: Bytes) -> io::Result<bool> {
+        match &mut self.inner {
+            DecoderInner::Identity(inner) => write_overflowing(inner, &bytes),
+            DecoderInner::Br(inner) => write_overflowing(inner, &bytes),
+            DecoderInner::Gzip(inner) => write_overflowing(inner, &bytes),
+            DecoderInner::Deflate(inner) => write_overflowing(inner, &bytes),
+        }
+    }
+
+    /// Returns decoded bytes from the Decoder's buffer.
+    ///
+    /// This can be called at any time during decoding. However, the limit stated during
+    /// initialization remains in effect across the entire decoded payload.
+    pub fn take(&mut self) -> Bytes {
+        match &mut self.inner {
+            DecoderInner::Identity(inner) => inner.take(),
+            DecoderInner::Br(inner) => inner.get_mut().take(),
+            DecoderInner::Gzip(inner) => inner.get_mut().take(),
+            DecoderInner::Deflate(inner) => inner.get_mut().take(),
+        }
+    }
+}
+
+impl fmt::Debug for Decoder {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.inner {
+            DecoderInner::Identity(inner) => f.debug_tuple("Identity").field(inner).finish(),
+            DecoderInner::Br(_inner) => f.debug_tuple("Br").finish(),
+            DecoderInner::Gzip(inner) => f.debug_tuple("Gzip").field(inner).finish(),
+            DecoderInner::Deflate(inner) => f.debug_tuple("Deflate").field(inner).finish(),
+        }
+    }
+}
+
+/// A payload based on [`SharedPayload`] that decompresses the request body on-the-fly.
+///
+/// Note that this stream is not fused.
+#[derive(Debug)]
+pub struct DecodingPayload {
```

Comment:
Convenient utility type for consuming streams of decoded chunks. However, in most cases we want to collect into a single buffer, in which case using the `Decoder` directly is more efficient. See `RequestBody` for this.